Item del

meltemozcan opened the pull request.

marklhc commented:
I suggest not including the `.gitignore` file in the merge.

For the vignettes, I'm not sure whether we can re-share the data files used in the vignette. Do we need to use the data files? An alternative would be to save the fitted `lavaan` object instead of the raw data.

marklhc commented:
And I think we only need one vignette, right?

marklhc commented:
I think we also need to add unit tests to the item deletion functions. This can be a separate issue.

meltemozcan commented:
> I suggest not including the `.gitignore` file in the merge.

I've deleted the .gitignore files now, they shouldn't be included in the merge but please let me know if there's an additional step I've missed.

> For the vignettes, I'm not sure whether we can re-share the data files used in the vignette. Do we need to use the data files? An alternative would be to save the fitted `lavaan` object instead of the raw data.
> And I think we only need one vignette, right?

item_deletion_examples.Rmd  is the only vignette we need, and does not need any of the data files but rather uses parameter estimates. I have deleted item_deletion_examples_other.Rmd and the data files.

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+#' @title
+#' Compute PS, SR, SE, SP weighted by group proportions
+#' 
```
It's a minor thing, but you can strip out the white spaces automatically if you use VS Code (Maybe RStudio as well.)

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+
+get_aggregate_CAI <- function(pmixr, store_summary) {
+  r <- store_summary$Reference; f <- store_summary$Focal
+  PS <- (pmixr*r[1] + (1 - pmixr)*f[1]) + (pmixr*r[2] + (1 - pmixr)*f[2])
```
How about first defining `pmixf = 1 - pmixr`?

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+#' excluded.
+
+err_improv_acai <- function(i, store_summary_full, store_summary_del1) {
+  r <- round(store_summary_full$Reference, 3)
```
Is it necessaryto do rounding here? Could you do rounding at the end instead?

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+  f_del1 <- round(store_summary_del1$Focal, 3)
+  if(# TP_f decreases/remains unchanged and TP_r increases, and either
+     # change has Cohen's h > 0.1
+    ((r[1] < r_del1[1]) & (f[1] >= f_del1[1]) & 
```
consider using `map()` or `mapply()` here.

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+  del_weight <- weights_item[del_i] # the weight to be redistributed
+   
+  # Unidimensional
+  if ((n_dim == 1) & (is.null(n_i_per_dim) | length(n_i_per_dim) == 1)) {
```
Consider using `&&` and `||`

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+#' @param del_i Index of the item to be deleted.
+#' 
+#' @return `new_w` Weights vector with redistributed weights.
+#' @examples
```
You can make these examples into testthat

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+#' @return `new_w` Weights vector with redistributed weights.
+#' @examples
+#' one_dim_w <- c(1:7)
+#' redistribute_weights(one_dim_w, del_i = 2)
```
It looks like the current redistribution works by redistributing the weight of the deleted item equally across items, regardless of their original weights. This isn't an issue when items have equal weights, but if some items have more weights than others, should the redistribution be proportional to the original weights?

meltemozcan replied:
You are right, I had originally written the function to assume that items in a given subscale would have equal weights. Thank you for bringing this to my attention, I will convert this to an issue and modify the redistribute_weights function.

meltemozcan replied:
Fixed in commit  6fcccf9a1dd9c9a6fafdb3c02e69add3121e8eab

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+#'                        nu_f = c(.225, -.05, .240, -.025, .125),
+#'                        Theta_r = diag(1, 5),  
+#'                        Theta_f = diag(c(1, .95, .80, .75, 1)), 
+#'                        weights = c(1/4, 1/4, 1/6, 1/6, 1/6))       
```
It seems a bit odd that weights constitute an input for determining biased items. Would it make sense to have the `setdiff` part outside of the function to determine which item is biased and still part of the scale?

marklhc commented on R/helper_item_deletion.R:
```diff
@@ -0,0 +1,341 @@
+  
+  # Compare factor loadings 
+  if(is.matrix(lambda_r)) {
+    for(i in seq_len(ncol(lambda_r))) {
```
Here's an alternative:

```r
which(!apply(lambda_r == lambda_f, 1, all))
```

marklhc replied:
And you can convert the input to matrix with `as.matrix(lambda_r)` to avoid if-else, although that may not be the most memory efficient.

YichiZhang2024 commented on DESCRIPTION:
```diff
@@ -1,6 +1,6 @@
 Title: Multidimensional Classification Accuracy Analysis Framework
-Version: 0.0.1.1
+Version: 0.0.2
 Authors@R: c(
```
We should add our names here!

marklhc replied:
I think our names are shown?

YichiZhang2024 commented on R/classes.R:
```diff
@@ -126,73 +120,74 @@ print.PartInvList <- function(x, ...) {
+    print(x$`PartInv by groups`$focal)#$outputlist[c(1, item_set+1)])
+    print(x$PartInv$strict)#$outputlist[c(1, item_set+1)])
+    print(x$PartInv$partial)#$outputlist[c(1, item_set+1)])
+  }
```
Just want to say this is really helpful for me to understand how the class works in R :)

YichiZhang2024 reviewed (approved):
Great job Meltem! I learned a lot from your codes and had my first real pull request! I don't have particular comments on the coding part, but if you need help with solving Mark's comments, I am happy to work on them together.

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+---
+title: "Applied examples"
```
Suggest to be more descriptive: e.g., "Applied examples of item deletion statistics"

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+The only biased item in the Positive Affect subscale is item 1. We see from the 
+first output table that deleting the first item increases $\ov{\SR}$ and 
+$\ov{\SE}$ from $0.883$ to $0.894$, and $\ov{\SP}$ from $0.752$ to $0.777$. The
+second output table shows that $h^{|1}\ov{\SR}=h^{|1}\ov{\SE}=-0.037$ and
```
What do you think about switching the sign for h, so that positive means improving? It may be helpful to check how alpha when item deleted was reported.

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+minimal-small effect size as shown in the final table. Removing any other item 
+harms CAI. 
+
+In light of the conflicting findings and the small effect size of the potential 
```
It'd be helpful to clarify what you mean by "conflicting findings." It may be worth pointing out we recompute the cut score?

meltemozcan replied:
Thank you, I modified the phrasing to address your first comment. 

I'm not sure what you mean by recomputing the cut score. Do you mean explicitly stating that each example (CESD POS, CESD DEP, CESD-full etc.) uses a different cut score, or are you referring to the drop-one example later on? As a side note, cut_z is currently 16 for both the full and the drop one examples. Do you think it makes sense to keep it 16, or round down to 15 from 16/60*(60-3)=15.2 (3 is the max score on each item, there are 20 items in the scale, and 16 is the cutoff used by Radloff on the full scale) @marklhc

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+---
+title: "Applied examples"
+output: rmarkdown::html_vignette
```
How about adding a table of content, given the large amount of information in the vignette?

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+ 
+Removing the second item in this 2-item subscale leads to lower $\ov{\CAI}$, and 
+the effect size of this decrease is $\h^{|2} \ov{\SR}=\h^{|2}\ov{\SE}=0.173$, which
+is approximately a small effect size. The adverse impact ratio goes up to $1$
```
I think we shouldn't use Cohen's benchmark to define what a small effect size is. This seems pretty big to me. Something we can further discuss.

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+using their domain specific expertise and consulting existing literature to 
+choose between the two items. The output object from item_deletion_h contains 
+other tables that may aid the researcher in this process. For instance, the 
+outputs of PartInv() may be examined for the two item deletion scenarios:
```
Change to `PartInv()`

marklhc commented on vignettes/item-deletion-examples.Rmd:
```diff
@@ -0,0 +1,584 @@
+own to make classification decisions.
+
+
+### Full 4-Factor CES-D Scale
```
Testing

marklhc commented on R/item_deletion_h.R:
```diff
@@ -0,0 +1,496 @@
+#'                                nu_f = c(.225, -.05, .240, -.025),
+#'                                Theta_r = diag(.96, 4),
+#'                                n_dim = 1, plot_contour = TRUE,
+#'                                print_formatted = TRUE)
```
Now that we use the `itemdeletion` class with a print method, I'm curious whether you think we should keep the `print_formatted` option? There are at least three options:

1. Keep it for both `item_deletion_h()` and `print.itemdeletion()`
2. Keep it only for `print.itemdeletion()`
3. Not keeping it anywhere.

marklhc commented on R/item_deletion_h.R:
```diff
@@ -0,0 +1,496 @@
+  
+   
+   rownames(delta_h_str_vs_par_ref) <- rownames(delta_h_str_vs_par_foc) <-
+     rownames(delta_h_R_vs_Ef) <-  c(paste0('\u0394', c("h(TP)")), 
```
Just an alternative

```r
paste0("\u0394h(", c("TP", "FP", "TN", "FN", "PS", "SR", "SE", "SP"), ")")
```

marklhc commented on R/item_deletion_h.R:
```diff
@@ -0,0 +1,496 @@
+  class(returned) <- "itemdeletion"
+  
+  return(returned)
+}
```
This doesn't need to be addressed in the pull request, but right now the function has 250 lines of code, and I feel like it may be helpful to move some of the components out as separate functions. For example, setting names and formatting.

marklhc reviewed (commented):
I think I've finished reviewing the files in the initial pull request. Let me know if you have any questions. Some of these can be addressed after the merge. Right now I don't see apparent issues with the merge.
